streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -517,16 +520,21 @@ private void restoreChangelog(final ChangelogMetadata changelogMetadata) {
         if (partitions.isEmpty())
             return Collections.emptyMap();
 
+        final Map<TopicPartition, Long> committedOffsets;
         try {
             // those do not have a committed offset would default to 0
-            return mainConsumer.committed(partitions).entrySet().stream()
+            committedOffsets =  mainConsumer.committed(partitions).entrySet().stream()
                 .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue() == null ? 0L : e.getValue().offset()));
         } catch (final TimeoutException e) {
             // if it timed out we just retry next time.
             return Collections.emptyMap();
         } catch (final KafkaException e) {
             throw new StreamsException(String.format("Failed to retrieve end offsets for %s", partitions), e);
         }
+
+        lastUpdateOffsetTime = time.milliseconds();
```

Comment:
If there's nothing to be updated or timed out, we do not update the timer.

Reply:
This would be less mysterious if this method were inlined into `updateLimitOffsets`. Right now, it's not terribly clear why it's ok to set the "last update offset time" in a method that doesn't update the offsets.

Reply:
This function is triggered by another caller besides `updateLimitOffsets`, plus it is a bit close to the NPathComplexity threshold..